### PR TITLE
Fix representation when printing abstract consts

### PR DESCRIPTION
**Pull request by w-utter**
Previously, when printing a const generic expr, it would only display it as `{{const expr}}`. This allows for a more legible representation when printing these out.

I also zipped the types with their constants for abstract consts that contain function calls when using type annotations, eg: `foo(S: usize, true: bool) -> usize` insteaad of `foo(S, true): fn(usize, bool) -> usize` for conciseness.

### CODE REVIEW

**Review by compiler-errors — APPROVED** [review]
I like this a lot

**Comment by compiler-errors** [comment]
Please squash this into one commit, then I can approve it.

@rustbot author

**Comment by jumbeldliam** [comment]
Thanks! Just squashed them into one commit.

I also want to look into formatting the statements with parentheses based on the expr tree when I get the chance, as something like `-(N + 1) * 2` prints as `-N + 1 * 2`, but couldn't seem to find a quick solution without wrapping every statement in parenthesis which I don't think looks nice

**Comment by compiler-errors** [comment]
Ah, I had totally overlooked that. Thanks for pointing it out. I... actually don't think we should be printing out *wrong* expressions, and we should fix this in this PR.

You should be able to get a precedence from a `hir::BinOp` (which is a `ast::BinOp`) by turning it first into an `AssocOp` via (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/util/parser/enum.AssocOp.html#method.from_ast_binop) then calling `precedence`/`fixity`. You can then compare the precedence and fixity of the LHS and RHS to know when to insert parentheses.

@rustbot author

**Comment by compiler-errors** [comment]
If you don't want to do this, let's just always parenthesize inner expressions, and I can implement a better heuristic for this.

**Comment by jumbeldliam** [comment]
No worries, this new commit should take precedence into account when adding parenthesis. 

I didn't bother with fixity as the unop kind does not have an equivalent `to_hir_binop()` and all fixities are already seperated by its respective expr kind so I just had a different implementation for each. 

Having a fn for precedence in `rustc_middle::ty::consts::kind::Expr` would be nice to have and would clean up some of the stuff I wrote, but it seemed out of scope for this pr

**Comment by jackh726** [comment]
r? compiler-errors

**Comment by Dylan-DPC** [comment]
@w-utter any updates on this? thanks

**Comment by w-utter** [comment]
I completely forgot about this after holiday break 😅 It's finished as far as I remember, I was just waiting on approval

Hopefully I can guess what the unconstrained generic const is first try and it can get merged as I don't have the environment to test it set up anymore

**Comment by Dylan-DPC** [comment]
@w-utter this still has a merge commit in the list of commits. You have to unmerge it or reset that commit using an interactive rebase.

**Comment by w-utter** [comment]
That took a lot longer than I'd like to admit but we should be all good to go now 😁

**Pull request merged**
